Implements part of #754 via getter/setters for proxying the file format 3 attributes

kontur opened the pull request:
This implements writing `postscriptBlueValues`, `postscriptOtherBlues`, `postscriptStemSnapV` and `postscriptStemSnapH` for glyphs > ufo.

I added [./tests/glyphs3_test.py](tests) and other tests are passing as well.

I'm not sure how elegant or clumsy it is to have those `GSFontMaster._...` underscore internals and accessing them via the getter/setter proxies, but it seemed the most practical way synthesize them from the `GSFont` attributes required for the offsets.

There are [this](https://github.com/kontur/glyphsLib/blob/fc2b551ddd02da4df0f0b2290fc526a554108847/Lib/glyphsLib/classes.py#L1765), [this](https://github.com/kontur/glyphsLib/blob/fc2b551ddd02da4df0f0b2290fc526a554108847/Lib/glyphsLib/classes.py#L1799) & [this](https://github.com/kontur/glyphsLib/blob/fc2b551ddd02da4df0f0b2290fc526a554108847/Lib/glyphsLib/classes.py#L1819) rather blunt stop gaps in the getters which mainly function to return nothing for cases where the masters are constructed via code, e.g. in the tests (about 50 or so fail otherwise). In a way this could mask a real issue as well, but not sure how realistic it is to rewrite the tests, or add some kind of "defaulting" structures to keep the indexes for constructed objects in sync between GSFont and GSFontMaster.

Should I squash my commits into one? The first commit attempts were a bit hit and miss...

kontur commented:
Any chance to get this reviewed? :)

anthrotype commented on tests/glyphs3_test.py:
```diff
@@ -22,3 +22,24 @@ def test_glyphspackage_load(datadir):
+    assert master.alignmentZones != []
+
+    # As per classes_tests::GSAlignmentZoneFromFileTest
+    for i, zone in enumerate([(800, 10), (700, 10), (470, 10), (0, -10), (-200, -10)]):
+        pos, size = zone
```
```suggestion
    for i, (pos, size) in enumerate([(800, 10), (700, 10), (470, 10), (0, -10), (-200, -10)]):
```

kontur replied:
Ah, nicer!

kontur commented on Lib/glyphsLib/classes.py:
```diff
@@ -1744,6 +1755,82 @@ def _set_metric(self, metricname, value):
+
+    @alignmentZones.setter
+    def alignmentZones(self, value):
+        assert type(value) == list
```
Yes, makes sense! It actually expects a list of tuples; I'll add tests for this, too.

kontur commented:
Rebased on `main` again, but there appear to be a lot of unrelated tests failing. What's the deal with those, are those residue from merged in WIP and the respective features will still fix those, or is there an effort to just tackle them as separate task?

anthrotype commented:
> there appear to be a lot of unrelated tests failing.

the CI checks seem to be passing now, not sure what you're referring to: https://github.com/googlefonts/glyphsLib/actions/runs/3627890015

I'll take another review pass

anthrotype commented:
the regression tests failing is expected when adding new features from a PR branch, but once that's merged to main, they fix themselves (the regression test CI job compares output from PR branch with that from main branch).

One needs to manually check that the changes are as expected:
https://github.com/googlefonts/glyphsLib/actions/runs/3627890014/jobs/6119377961

anthrotype commented:
the CI also runs tox the same way as you do so that's unexpected.. let me try reproduce locally from a copy of your PR branch

anthrotype commented:
the regression tests failures seem all benign, I can see that postscriptBlueValues are now being compiled which is what this PR is about

kontur commented:
Ah, perfect, thanks for the `-r` hint, not very familiar with tox. Tests run locally without issues as well 🙇

anthrotype commented on Lib/glyphsLib/classes.py:
```diff
@@ -1747,6 +1760,91 @@ def _set_metric(self, metricname, value):
+        )
+
+        zones = []
+        for index, fontMetric in enumerate(self.font.metrics):
+            if fontMetric.type not in validZoneMetrics:
+                continue
+            metric = self.metrics[index]
```
I'm a bit confused.. Where are the metrics (or the stems in a similar piece of code below) actually stored? Inside the GSFont.metrics or in this GSFontMaster.metrics?
Why do you have to iterate over the font.metrics if you then take the item from the master's metrics?

anthrotype replied:
ah ok it's explained in https://github.com/schriftgestalt/GlyphsSDK/blob/Glyphs3/GlyphsFileFormat/GlyphsFileFormatv3.md#differences-between-version-2

anthrotype commented on Lib/glyphsLib/classes.py:
```diff
@@ -1747,6 +1760,91 @@ def _set_metric(self, metricname, value):
+
+    @alignmentZones.setter
+    def alignmentZones(self, entries):
+        if type(entries) is not list:
+            raise ValueError(
```
usually it's preferable to use isinstance instead of `type() is ...`, unless you really have a good reason to restrict the exact type. And raise TypeError instead of ValueError when the type is not as expected

kontur commented:
@anthrotype yep, those type checks are better now, thanks for pointing that out.

regarding @schriftgestalt 's suggestion: I agree in general, but this would probably need a more orchestrated effort overall, no? it shouldn't just prioritise v3 handling for one attribute. I can rework this for the `alignmentZones` if that will otherwise prevent the PR from being merged. personally I just need to be able to work with v3 files, from releases preferably.

anthrotype commented on Lib/glyphsLib/classes.py:
```diff
@@ -1444,6 +1444,11 @@ def __repr__(self):
         return (self.position, self.size) < (other.position, other.size)
 
+    def __eq__(self, other):
+        if not hasattr(other, "position") or not hasattr(other, "size"):
+            return False
```
ping @kontur

kontur replied:
👍 missed this one before

anthrotype commented:
does glyphs.app python API still provide getters like `alignmentZones` returning a list of [GSAlignmentZone](https://docu.glyphsapp.com/#GSAlignmentZone) objects for v3 as well as v2 fonts? If so, we could keep the old v2 API and file an issue to expose the new v3 API alongside it.

simoncozens commented:
I said the other day to Cosimo that we should just support the G3 model and have a separate bit of code to “upgrade” G2 files on load. Trying to manipulate both systems is making the code atrocious, and G3 is much cleaner. But this would be a major overhaul.

anthrotype commented on Lib/glyphsLib/classes.py:
```diff
@@ -1574,30 +1581,36 @@ def _serialize_to_plist(self, writer):
         "ascender": 800,
         "descender": -200,
         "italic angle": 0,
+        "baseline": 0,
```
what does this change have to do with this PR? what is it meant to fix?

anthrotype commented on Lib/glyphsLib/classes.py:
```diff
@@ -1747,6 +1760,91 @@ def _set_metric(self, metricname, value):
+            "cap height",
+            "x-height",
+            "baseline",
+            "descender",
```
where's this list of validZoneMetrics taken from? it seems constant, it should probably be defined at global scope, or in the constants module.

anthrotype replied:
and why skip some of the font metric types as not "valid" when fetching a given master metric values? where is this logic encoded?

kontur replied:
Basically I went by what's in the UI (mind you I missed `bodyHeight`):

<img width="501" alt="Screenshot 2022-12-16 at 21 09 41" src="https://user-images.githubusercontent.com/1893577/208174153-30b8b9ea-26ac-467e-9d4c-d9571ceb57a2.png">

Outputting the `GSMetric.type` of these objects of a file gives:

```
ascender
cap height
x-height
baseline
descender
italic angle
bodyHeight

```

The last empty line for type is what the UI has as "Custom" metric. This may or may not be an overshoot — presumably things with overshoot != 0 we want to treat as an alignmentZone as well?

I think when I did a "has no overshoot" exclusion instead of explicitly checking for the type _a lot_ of tests outright failed because the font metricValues does not match a given master's metricValues length (e.g. in all sorts of constructed test cases or programmatic access of those objects that don't adhere to the file format 3 referencing the font metricValues from the master by index). As far as I recall the `"baseline": 0` default above also addressed this.

Note sure what to do here, but I'll ponder it.

kontur replied:
Updated in code below in forced push. Disregarding the "italic angle" type of metrics early was key to avoiding all those issues in tests that construct just masters. Gets rid of the explicit magic list and does not require the `baseline` default either. Thanks for poking holes @anthrotype :)

anthrotype replied:
> ... because the font metricValues does not match a given master's metricValues length (e.g. in all sorts of constructed test cases or programmatic access of those objects that don't adhere to the file format 3 referencing the font metricValues from the master by index

if it's invalid that len(master.metricValues) != len(font.metrics) then it is those tests that would need to be fixed...

kontur replied:
The tests were correct, only when _not filtering_ things that are now returned by a master'a metrics which are not alignmentZones in format 3 those test broke because without knowledge of the master my previous iteration of the code could not filter those properly. Different approach in the updated code, essentially discarding the `italic angle` from being an alignmentZone outright, and further removing metrics which do not have an overshoot, e.g. do not form a bluezone because they would be a "line" only.

anthrotype replied:
can a valid "italic angle" metric ever have overshoot != 0? I suppose not. And if not, then simply filtering out metrics with overshoot=0 (without special-casing italic angle itself) should accomplish the same thing.

kontur replied:
In
```
 for index, fontMetric in enumerate(self.font.metrics):
            # Ignore the "italic angle" "metric", it is not an alignmentZone
            if fontMetric.type == "italic angle":
                continue
            metric = self.metrics[index]
            # Ignore metric without overshoot, it is not an alignmentZone
            if metric.overshoot == 0:
                continue
            zone = GSAlignmentZone(pos=metric.position, size=metric.overshoot)
            zones.append(zone)
```

`self.font.metrics` is a [GSMetric](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/Lib/glyphsLib/classes.py#L1393) value which has no overshoot. One has to use the index of this  _font_ metric values to get the _master_ [GSMetricValue](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/Lib/glyphsLib/classes.py#L1408) value, and then discard things which have no overshoot. Filtering the _font_ metric values by type `italic angle` is necessary because just accessing `master.metrics[]` by index will not have a value for a _font_ metric that is an italic angle.

The format 3 handling of metric values referencing back up to the font metrics list is convoluted imo.

anthrotype commented on Lib/glyphsLib/classes.py:
```diff
@@ -1747,6 +1760,91 @@ def _set_metric(self, metricname, value):
+        return zones
+
+    @alignmentZones.setter
+    def alignmentZones(self, entries):
```
looks like alignmentZones is a read-only (https://docu.glyphsapp.com/#GSFontMaster.alignmentZones) so maybe we should not add a setter.
I agree with @simoncozens when he wrote

> We should store things as v3 format internally to the object but provide accessors to get at information in a v2 way.

kontur replied:
Some of the existing tests assign `alignmentZones` like this for constructing objects via code only (as opposed to actual conversion). Happy to drop the setter and tests.

anthrotype replied:
If one wants to _set_ these alignment zones using Glyphs.app v3 scripting API (where alignmentZones is read-only), one would one do that exactly?
Does glyphsLib already support this? If not, what do we need to support it?

kontur replied:
Going by what I found in the code. The primary use case in the library seems to be [here](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/Lib/glyphsLib/builder/blue_values.py#L55) setting the alignmentZones when converting from UFO to glyphs. 

Also tests [here](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/tests/classes_test.py#L755), [here](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/tests/writer_test.py#L301), [here](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/tests/builder/builder_test.py#L495) and [here](https://github.com/googlefonts/glyphsLib/blob/74aefb387a5a84659902b53fa805b74ff3d8ed0c/tests/builder/builder_test.py#L504) — my assumption being that a new feature or refactor should not break existing tests and thus a setter needs to be supported.

anthrotype replied:
i'm ok to keep a setter for alignmentZones, but I wonder maybe it should store them internally as v3 metricValues.
Say you go from UFO => glyphs and want to save as format_version=3, those zones will not be written out in the generated .glyphs file.
I think we should have a method to convert from v2 (alignmentZones and stems) to v3 metricValues and call that when saving a font and format_version=3.

anthrotype replied:
but that setter only sets `self._alignmentZones`, which as far as I understand is _not_ written out to .glyphs source when format_version=3,

```python
        if writer.format_version == 2:
            writer.writeObjectKeyValue(self, "alignmentZones", "if_true")
```

anthrotype replied:
so I presume that roundtripped glyphs file of yours has format version 2.. but when Glyphs.app v3 opens it up it automatically populate the metric values.
That's what we are missing here, a way to upgrade these metrics from v2 to v3. I am not saying that we should do within this same PR, just that this is something we would like to support

kontur replied:
Ah, I see, didn't understand that's what's happening. Yes, correct. 

I'll have a look at that, either as part of this or separate. Follows the same obtuse logic of font metrics vs master metric 😄 Happy to adopt that if there is a more general v2 > v3 structure in place. That section of code you quoted, I suppose, is at the heart of what would need a more elegant and transparent file format version upgrade.

kontur replied:
_How_ does one actually write something out in format 3? Tried to find tests that verify those other `self.format == 3` written things (e.g. [here onwards](https://github.com/googlefonts/glyphsLib/blob/24b4d340e4c82948ba121dcfe563c1450a8e69c9/Lib/glyphsLib/classes.py#L4234)), but could only find [this one](https://github.com/googlefonts/glyphsLib/blob/24b4d340e4c82948ba121dcfe563c1450a8e69c9/tests/writer_test.py#L1190). Am I completely off track or is there pretty little actually for outputting/writing/testing v3? Would testing the output of the writer in `writer_tests.py` the right spot to implement checking for a yet-to-be-done implementation of writing those metrics back to v3 files?

schriftgestalt replied:
> but could only find [this one](https://github.com/googlefonts/glyphsLib/blob/24b4d340e4c82948ba121dcfe563c1450a8e69c9/tests/writer_test.py#L1190).

This looks like the way to do this. While writing out, check the format and what data you have and act accordingly.

kontur commented:
Rebased and refactored the `alignmentZones` getter code filtering that determined what Metric is or is not an alignmentZone.

anthrotype commented on Lib/glyphsLib/classes.py:
```diff
@@ -1747,6 +1759,87 @@ def _set_metric(self, metricname, value):
+            metric = self.metrics[index]
+            # Ignore metric without overshoot, it is not an alignmentZone
+            if metric.overshoot == 0:
+                continue
```
can you confirm that Glyphs.app v3 also does the same (ignore 'italic angle' metric an those with overshoot=0) when you get a GSFont.alignmentZones?
What's an example of such a metric without overshoot?

schriftgestalt replied:
Yes. This is what Glyphs is doing.

kontur commented:
@anthrotype Ah sorry, must have missed the lint issues locally and didn't follow the updates of the test runner here. Fixed and rebased on `main` once more, gtg from my part.

kontur commented:
*patiently waiting*

anthrotype commented:
thanks for the patience
